### ISSUE TITLE
💡 [Feature]: Add possibility to add custom additional step

### ISSUE THREAD

**Issue by Adam-it** [pnp/vscode-viva#217]
### 🎯 Aim of the feature

We should add some possibilities, preferably extending it as a setting for the extension, to allow devs to define their own additional setup options. 
We could expose a setting in which users could manage a json array that will have a `label` and `command`. `label` will hold the string that will be present in the UI in the scaffolding form and `command` will be the npm command that should be executed after this option was selected.

example idea for a this kind of json object:
```json
[
  {
    "label": "PnP JS SharePoint",
    "command": "npm install @pnp/sp"
  }
]
```

We should NOT allow to disable the pre-defined options

### 🤔 Additional remarks or comments

doing this change we should keep an eye out for this issue #216
if it is already done we should also add a option like `visiblity` that will allow to list out in as string array components that should have this option visible

**Comment by ervingayle** [comment]
Hi @Adam-it. How did you envision the array being stored?

**Comment by ervingayle** [comment]
Hi @Adam-it . Thank you for the example. This helps me. I did not know of this option. I know that you mentioned not allowing for removal of the default ones but this json array would only specify the additional packages that someone would want installed. Yes, I believe that this addition would require good docs with examples to effectively help everyone.

**Comment by ervingayle** [comment]
Yes, please assign this to me.

**Comment by Adam-it** [comment]
> Yes, please assign this to me.

All yours 👍

**Comment by ervingayle** [comment]
Hi @Adam-it. I have made progress with this item but I do have some issues with the implementation. Currently I tried with selecting React and React Dom, PnP and then I added a new "Additional Step" called Install custom steps. This reads the setting value and then I loop through the json. This works. However, what I am seeing is that the React and React Dom step is being bundled with the install dependencies initial powershell task where npm i is executed. Every additional custom step is created in a new terminal window and installed simultaneously. In some cases the package is installed and no errors are shown, folder in node_modules but the package entry is not in package.json. I have tested this about 15 times so far. Do you know what would cause this?

Also, I intended to show what would be installed in the scaffolding form but passing the setting value to the view didn't work as I expected so for now I have reverted that code.
I have not opened a pull request as yet. Code [here](https://github.com/ervingayle/vscode-viva/)

**Comment by Adam-it** [comment]
@ervingayle thanks for reaching out. 
So I am not 100% if I understand everything you wrote but I think I kinda get what the problem is. From what I understand the code you added that should loop over the custom additional steps simply fires in parralel rather than sequentially and that is why you see multiple terminal opening with multiple `npm i...` commands running. 

I checked your commits and most probably I would investigate this part
https://github.com/ervingayle/vscode-viva/blob/74c58c5de15c4a6026739de35cd7b69fae2912e8/src/extension.ts#L96-L100

Here you have a `forEach` that esentially returns multiple async functions and each of those run in parallel. It is not a loop where every iteration of the loop awaits to finish until the next iteration is taken. I would suggest to refactor it to a `for` loop and reacheck.

Let me know if it helps.

**Comment by ervingayle** [comment]
Hi @Adam-it . Thank you for the response. I tried with a for loop and get similar behavior. I realize the behavior that I am seeing is due to the fact that I set a unique title for each package that is set in the extension setting and the TerminalCommandExecutor is looking at title/name value to determine if a new terminal should be created. I don't change the name for the React, React Dom install option but I did for the custom steps which is why I was confused at first. Its tested and working now.

**Comment by ervingayle** [comment]
Hi @Adam-it . Thanks for your help. PR submitted for this item.

**Comment by Adam-it** [comment]
> Hi [@Adam-it](https://github.com/Adam-it) . Thanks for your help. PR submitted for this item.

You Rock 🤩

**Comment by ervingayle** [comment]
Hi @Adam-it . Thank you for all your help. The guidance you provided helped. I have created a new PR for this issue.

**Comment by Adam-it** [comment]
@nicodecleyre, @Saurabh7019, what about this?
In general, I think it is (was) a good idea, but:
- The way we would prepare it is kinda hacky - having json as text in VS Code extension settings. This means this feature would be really advanced and I wonder if anyone would actually use it
- Most probably, scaffolding will need to change anyway as MS is moving away from yeoman. So this feature would be a thing of a past. 
- as we want to add support for all SPFx versions this would become even more tricky. I would leave it out for now and get back to it once we have scaffolding form for all SPFx versions done

**Comment by Saurabh7019** [comment]
> * I wonder if anyone would actually use it

I agree.. I don't think this feature is really needed. The chances of developers extending the scaffolding form this way, instead of simply installing the packages once the solution is created, seem very low. To be honest, I don't see myself using it either.
